### ISSUE TITLE
Type check of dict content fails

### ISSUE THREAD

**Issue by Zaikaba** [microsoft/pyright#9111]
## Environment data

-   Language Server version: 2024.9.2
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 3.12.5 ( conda )
-   python.analysis.indexing: true
-   python.analysis.typeCheckingMode: strict

## Code Snippet

```python
# Type of "bar1" is partially unknown
#   Type of "bar1" is "dict[str, Unknown]"
bar1 = { 
    "a": True, 
    "b": "meow"
}

# fine: "bar2: dict[str, bool | str]"
bar2 = dict( { 
    "a": True, 
    "b": "meow"
} )
```

## Expected behavior

The type checker should derive the type of bar1 and bar 2 as dict[str, bool | str] ( or perhaps dict[str, Any] if there are too many types ).

## Actual behavior

Type derivation fails for bar1 ( as indicated in code ) and shows an error, succeeds, however, when wrapped into a dict() call, as in bar2.

## Logs

```

```

**Comment by erictraut** [comment]
Pyright is working as intended here. Refer to [this documentation](https://microsoft.github.io/pyright/#/type-inference?id=dictionary-expressions) for information about how pyright infers the type of dictionary expressions.

In your second code sample, you are calling the `dict` constructor, and the type of this constructor call is dictated by typeshed type definitions in `builtins.pyi`.
